issue #1393 Create validator for typed value

IvanPizhenko opened the pull request:
Second attempt of solving issue #1393. This time without extending class hierarchy.

IvanPizhenko commented:
@tipsy So what do you think of this second variant?

tipsy commented:
Hey @IvanPizhenko, I'm a bit busy lately, I will have a look in the weekend. From a first glance it seemed to not be backwards compatible?

IvanPizhenko commented:
@tipsy ok, thank you. It should be compatible with what you have now (i.e. in 4.0.0), because there is additional constructor that replicates current one. And you can see, I did not touch any existing tests, they all compile and pass, just added new ones to cover new usage scenario.

tipsy commented:
I probably missed something, I'll have a closer look this weekend :)

IvanPizhenko commented:
@tipsy So did you check it? You told me you'll look on the last weekend... I kindly ask you, please consider reviewing this faster. I'm porting application from Javalin 3 to Javalin 4 and I've submitted this to get this faster into Javalin.

tipsy commented on javalin/src/main/java/io/javalin/core/validation/BaseValidator.kt:
```diff
@@ -57,11 +92,17 @@ open class BaseValidator<T>(val stringValue: String?, val clazz: Class<T>, val f
         return this
     }
 
+    constructor(
```
Could you move this up?

tipsy commented on javalin/src/main/java/io/javalin/core/validation/Validator.kt:
```diff
@@ -14,10 +14,17 @@ import org.jetbrains.annotations.NotNull
  * The [check] method wraps its non-nullable predicate in a nullable predicate
  */
-open class Validator<T>(stringValue: String?, clazz: Class<T>, fieldName: String) : BaseValidator<T>(stringValue, clazz, fieldName) {
+open class Validator<T>(
```
Could you restore the old formatting?

tipsy commented on javalin/src/main/java/io/javalin/core/validation/NullableValidator.kt:
```diff
@@ -6,7 +6,13 @@
 package io.javalin.core.validation
 
-open class NullableValidator<T>(stringValue: String?, clazz: Class<T>, fieldName: String) : BaseValidator<T>(stringValue, clazz, fieldName) {
+open class NullableValidator<T>(
```
Could you restore the old formatting?

tipsy commented on javalin/src/main/java/io/javalin/core/validation/Validator.kt:
```diff
@@ -27,9 +34,8 @@ open class Validator<T>(stringValue: String?, clazz: Class<T>, fieldName: String
-        stringValue == null -> default
-        else -> super.get()!!
+    fun getOrDefault(default: T): T {
+        return if (stringSource?.stringValue != null || typedValue != null) super.get()!! else default
```
Could `stringSource?.stringValue != null || typedValue != null` become a function?

tipsy commented on javalin/src/main/java/io/javalin/core/validation/BaseValidator.kt:
```diff
@@ -15,28 +15,63 @@ data class Rule<T>(val fieldName: String, val check: Check<T?>, val error: Valid
-            }
-            if (this !is NullableValidator && typedValue == null) { // only check typedValue - null might map to 0, which could be valid?
-                return@lazy mapOf(fieldName to listOf(ValidationError("NULLCHECK_FAILED", value = stringValue)))
+        if (stringSource != null) {
```
Could you restore the original formatting, so it's easier to see what has changed? Is it just moved inside an `if`, or did something else change?

tipsy commented:
> because there is additional constructor that replicates current one. 

Yeah, I didn't see that the first time around.

IvanPizhenko commented:
@tipsy Tried my best to solve above remarks. Please recheck.

tipsy commented:
Thanks @IvanPizhenko, it was a lot easier to review now :)

IvanPizhenko commented:
@tipsy Thank you! Now can you please post a new version which includes this one to the Maven Central?

tipsy commented:
If you need it right away I suggest using https://jitpack.io/ or similar tools.

IvanPizhenko commented:
I need it exactly in the Maven Central.

IvanPizhenko commented:
But thank you for pointing me to jitpack, didn't know such service ever exists.

IvanPizhenko commented:
@tipsy Any ideas when this will get published to the Maven Central?

tipsy commented:
@IvanPizhenko I usually release in the weekends, but I like to wait for more than 1 feature before creating a new release.

IvanPizhenko commented:
So when? Next weekend?

IvanPizhenko commented:
Generally, even one feature is good enough to release, especially if people wanted it and even agreed to contribute themselves.

pkkummermo commented:
Tipsy has a release regime that he usually stick to, and he offered a solution through using jitpack if this was critical for your application. It will be released, but give the maintainer some leeway in how he conducts the project and release lifecycle :)

IvanPizhenko commented:
@pkkummermo Unfortunately, jitpack just doesn't work.
I've read instructions presented on the jitpack.io main page and then read its more detailed docs, and then I've added to my `build.gradle.kts` jitpack repo:
```kotlin
repositories {
  maven("https://jitpack.io")
}
```
and changed library references from
```kotlin
api("io.javalin:javalin:4.0.1")
api("io.javalin:javalin-openapi:4.0.1")
```
to
```kotlin
api("com.github.tipsy.javalin:javalin:master-SNAPSHOT")
api("com.github.tipsy.javalin:javalin-openapi:master-SNAPSHOT")
```
as it was recommended on the jitpack docs, then tried to build the project and received these errors:
```
Could not resolve com.github.tipsy.javalin:javalin:master-SNAPSHOT
Could not resolve com.github.tipsy.javalin-openapi:javalin:master-SNAPSHOT
```
What am I doing wrong?

pkkummermo commented:
I have no experience with gradle, but that looks correct to me as per [the gradle instructions](https://jitpack.io/#tipsy/javalin/master-SNAPSHOT). You can also try to set it to a hash and see if it works. It needs to build the jar first, so there might also be some delays between the first initial query and the following ones. If you're depending on a master snapshot you also need to consider resolution strategies.

IvanPizhenko commented:
> 
> 
> I have no experience with gradle, but that looks correct to me as per [the gradle instructions](https://jitpack.io/#tipsy/javalin/master-SNAPSHOT). You can also try to [set it to a hash](https://jitpack.io/#tipsy/javalin/cd9ad16aa4) and see if it works. It needs to build the jar first, so there might also be some delays between the first initial query and the following ones. If you're depending on a master snapshot you also need to consider resolution strategies.

@pkkummermo  Sure, I've tried using the commit hash too, with the same no luck.

tipsy commented:
@IvanPizhenko https://javalin.io/news/2021/10/02/javalin-4.1.0-released.html

IvanPizhenko commented:
@tipsy Thank you!
